src/main/java/org/apache/mesos/hdfs/executor/NameNodeExecutor.java

```diff
@@ -105,14 +112,106 @@ public void frameworkMessage(ExecutorDriver driver, byte[] msg) {
+          PreNNInitTask checker = new PreNNInitTask(driver, nameNodeTask);
+          timer.scheduleAtFixedRate(checker, 0, 15000);
+        } else if (messageStr.equals(HDFSConstants.NAME_NODE_BOOTSTRAP_MESSAGE)) {
+          PreNNBootstrapTask checker = new PreNNBootstrapTask(driver, nameNodeTask);
+          timer.scheduleAtFixedRate(checker, 0, 15000);
+        }
+      } else {
         runCommand(driver, nameNodeTask, "bin/hdfs-mesos-namenode " + messageStr);
         startProcess(driver, nameNodeTask);
-        startProcess(driver, zkfcNodeTask);
-        driver.sendStatusUpdate(TaskStatus.newBuilder()
-            .setTaskId(nameNodeTask.taskInfo.getTaskId())
-            .setState(TaskState.TASK_RUNNING)
-            .setMessage(messageStr)
-            .build());
+      }
+    }
+  }
+
+  /**
+   * This class is designed to help launch the primary namenode in its init phase. The node needs
+   * the journal nodes to be available on port 8485. In Mesos-DNS, this takes an uncertain amount of
+   * time, so we must schedule it asynchronously.
+   */
+  private class PreNNInitTask extends TimerTask {
+    ExecutorDriver driver;
+    Task task;
+
+    public PreNNInitTask(ExecutorDriver driver, Task task) {
+      this.driver = driver;
+      this.task = task;
+    }
+
+    @Override
+    public void run() {
+      boolean success = true;
+      for (int i = schedulerConf.getJournalNodeCount(); i > 0; i--) {
+        String host = HDFSConstants.JOURNAL_NODE_ID + i + "." + schedulerConf.getFrameworkName() + "." + schedulerConf.getMesosDnsDomain();
+        log.info("Checking for " + host);
+        try (Socket connected = new Socket(host, HDFSConstants.JOURNAL_NODE_HEALTH_PORT)) {
+          log.info("Successfully found " + host + " at port " + HDFSConstants.JOURNAL_NODE_HEALTH_PORT);
+        } catch (SecurityException | IOException e) {
+          log.info("Couldn't resolve host " + host + " at port " + HDFSConstants.JOURNAL_NODE_HEALTH_PORT);
+          success = false;
+          break;
+        }
+      }
+      if (success) {
+        log.info("Successfully found all nodes needed to continue. Sending message: " + HDFSConstants.NAME_NODE_INIT_MESSAGE);
+        this.cancel();
+        runCommand(driver, task, "bin/hdfs-mesos-namenode " + HDFSConstants.NAME_NODE_INIT_MESSAGE);
+        startProcess(driver, task);
+      }
+    }
+  }
+
+  /**
+   * This class is designed to help launch the secondary namenode in its bootstrap phase. The node
+   * needs both name nodes to be available as DNS entries. In Mesos-DNS, this takes an uncertain
+   * amount of time, so we must schedule it asynchronously.
+   */
+  private class PreNNBootstrapTask extends TimerTask {
+    ExecutorDriver driver;
+    Task task;
+
+    public PreNNBootstrapTask(ExecutorDriver driver, Task task) {
+      this.driver = driver;
+      this.task = task;
+    }
+
+    @Override
+    public void run() {
```

Comment:
Perhaps as part of the simplification of this class, you can instead perform this check as a condition for the scheduler to launch another node. That way if mesos dns option is selected, you also check whether it has registered with Mesos DNS or not before moving to the next state in the state machine. That would be a  less hacky way than making the code changes in this class .